cloud-node-lifecycle controller: add fallback for empty providerID in shutdown

mkowalski opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Simiarly to the function `ensureNodeExistsByProviderID`, `shutdownInCloudProvider` should have a logic where in case of an empty providerID we get it using the name of the node. This is to support scenarios when the function is called with Node object that has a name but does not have any provider ID.

Currently in such a scenario we have an error as it is not possible to call `InstanceShutdownByProviderID` with empty value. With this change in such a scenario we will first obtain a correct provider ID and only afterwards check the shutdown status.

#### Which issue(s) this PR fixes:

Fixes: https://github.com/kubernetes/kubernetes/issues/120617

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

mkowalski committed:
cloud-node-lifecycle controller: add fallback for empty providerID in shutdown

Simiarly to the function `ensureNodeExistsByProviderID`,
`shutdownInCloudProvider` should have a logic where in case of an empty
providerID we get it using the name of the node. This is to support
scenarios when the function is called with Node object that has a name
but does not have any provider ID.

Currently in such a scenario we have an error as it is not possible to
call `InstanceShutdownByProviderID` with empty value. With this change
in such a scenario we will first obtain a correct provider ID and only
afterwards check the shutdown status.

Signed-off-by: Mat Kowalski <mko@redhat.com>

dims commented:
/ok-to-test

aojea commented:
we should add an unit test to cover regressions, please

mkowalski commented on staging/src/k8s.io/cloud-provider/fake/fake.go:
```diff
@@ -312,6 +312,11 @@ func (f *Cloud) InstanceExistsByProviderID(ctx context.Context, providerID strin
 func (f *Cloud) InstanceShutdownByProviderID(ctx context.Context, providerID string) (bool, error) {
 	f.addCall("instance-shutdown-by-provider-id")
+
+	if providerID == "" {
+		return false, fmt.Errorf("cannot shutdown instance with empty providerID")
+	}
```
This is not strictly needed here, but it's to show that the previous version of the code would have failed the unit test I'm adding right now. I could drop it because it's no-op now, but for the sake of avoiding regressions in the future it may be worth keeping

aojea replied:
if that mimics that mimics the cloud-providers implementations it seems ok, is this a general behavior?

mkowalski replied:
Today when we call `shutdownInCloudProvider` with empty providerID it blindly calls `InstanceShutdownByProviderID` passing the empty ID. The former exists only in Azure, GCE and vSphere provider.

In Azure for empty providerID we return `false, nil` (so no error thrown).

In GCP the function is not implemented and always returns `false, cloudprovider.NotImplemented` (so technically an error)

In vSphere the function immediately calls `GetNodeNameFromProviderID(providerID)` and that one for empty argument will return an error `Error while getting nodename for providerID none` (which is somehow misleading).

So overall the behaviour is inconsistent but in majority of cases (2 out of 3) we error if try to shutdown a nil-instance.

mkowalski commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

A flake(?)/timeout in `[It] [sig-api-machinery] Watchers should observe an object deletion if it stops meeting the requirements of the selector [Conformance] expand_less`

aojea commented:
yeah, it looks an unrelated failure

overall lgtm, it seems there is a possibility of saving one cloudprovide api call when the providerID is missing, but I'm not familiar enough with this controller 😅 
/assign @andrewsykim @wojtek-t

bridgetkromhout commented:
/assign @kmala @calebikhuohon
/triage accepted

kmala commented on staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -207,12 +208,24 @@ func shutdownInCloudProvider(ctx context.Context, cloud cloudprovider.Interface,
-	if err == cloudprovider.NotImplemented {
-		return false, nil
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
```
having additional check is fine but for my understanding may i know why the node doesn't have provider id ? Is it because the node is registered manually? Because kubelet seems to make sure provider id is set https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L388-L393

mkowalski replied:
I am by no means close to explain why that happens, but I dumped the whole Node object passed around and I can only confirm that for some reason it does not have provider id in the spec

```
&Node{ObjectMeta:{ci-ln-1fllgz2-c1627-c6gh5-master-0    178d7133-fb6d-4550-a575-f0bcbc0f430d 4505 0 2023-09-14 12:05:36 +0000 UTC <nil> <nil> map[beta.kubernetes.io/arch:amd64 beta.kubernetes.io/os:linux kubernetes.io/arch:amd64 kubernetes.io/hostname:ci-ln-1fllgz2-c1627-c6gh5-master-0 kubernetes.io/os:linux node-role.kubernetes.io/control-plane: node-role.kubernetes.io/master: node.openshift.io/os_id:rhcos] map[alpha.kubernetes.io/provided-node-ip:192.168.130.10,fd65:a1a8:60ad:271c::5a volumes.kubernetes.io/controller-managed-attach-detach:XXXX] [] [] []},Spec:NodeSpec{PodCIDR:,DoNotUseExternalID:,ProviderID:,Unschedulable:false,Taints:[]Taint{Taint{Key:node-role.kubernetes.io/master,Value:,Effect:NoSchedule,TimeAdded:<nil>,},Taint{Key:node.cloudprovider.kubernetes.io/uninitialized,Value:XXXX,Effect:NoSchedule,TimeAdded:<nil>,},Taint{Key:node.kubernetes.io/not-ready,Value:,Effect:NoSchedule,TimeAdded:<nil>,},},ConfigSource:nil,PodCIDRs:[],},Status:NodeStatus{Capacity:ResourceList{cpu: {{4 0} {<nil>} 4 DecimalSI},ephemeral-storage: {{128248430592 0} {<nil>}  BinarySI},hugepages-1Gi: {{0 0} {<nil>} 0 DecimalSI},hugepages-2Mi: {{0 0} {<nil>} 0 DecimalSI},memory: {{16759611392 0} {<nil>}  BinarySI},pods: {{250 0} {<nil>} 250 DecimalSI},},Allocatable:ResourceList{cpu: {{3500 -3} {<nil>} 3500m DecimalSI},ephemeral-storage: {{114349845518 0} {<nil>} 114349845518 DecimalSI},hugepages-1Gi: {{0 0} {<nil>} 0 DecimalSI},hugepages-2Mi: {{0 0} {<nil>} 0 DecimalSI},memory: {{15581011968 0} {<nil>}  BinarySI},pods: {{250 0} {<nil>} 250 DecimalSI},},Phase:,Conditions:[]NodeCondition{NodeCondition{Type:MemoryPressure,Status:False,LastHeartbeatTime:2023-09-14 12:11:02 +0000 UTC,LastTransitionTime:2023-09-14 12:05:36 +0000 UTC,Reason:KubeletHasSufficientMemory,Message:kubelet has sufficient memory available,},NodeCondition{Type:DiskPressure,Status:False,LastHeartbeatTime:2023-09-14 12:11:02 +0000 UTC,LastTransitionTime:2023-09-14 12:05:36 +0000 UTC,Reason:KubeletHasNoDiskPressure,Message:kubelet has no disk pressure,},NodeCondition{Type:PIDPressure,Status:False,LastHeartbeatTime:2023-09-14 12:11:02 +0000 UTC,LastTransitionTime:2023-09-14 12:05:36 +0000 UTC,Reason:KubeletHasSufficientPID,Message:kubelet has sufficient PID available,},NodeCondition{Type:Ready,Status:False,LastHeartbeatTime:2023-09-14 12:11:02 +0000 UTC,LastTransitionTime:2023-09-14 12:05:36 +0000 UTC,Reason:KubeletNotReady,Message:container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:Network plugin returns error: No CNI configuration file in /etc/kubernetes/cni/net.d/. Has your network provider started?,},},Addresses:[]NodeAddress{NodeAddress{Type:InternalIP,Address:192.168.130.10,},NodeAddress{Type:InternalIP,Address:fd65:a1a8:60ad:271c::5a,},NodeAddress{Type:Hostname,Address:ci-ln-1fllgz2-c1627-c6gh5-master-0,},},DaemonEndpoints:NodeDaemonEndpoints{KubeletEndpoint:DaemonEndpoint{Port:10250,},},NodeInfo:NodeSystemInfo{MachineID:3c688e98dbd040e8828076dcd2aab565,SystemUUID:94481b42-1446-7993-f130-543b246a5a97,BootID:af3f4814-10a0-4efa-979e-610779962d55,KernelVersion:5.14.0-284.32.1.el9_2.x86_64,OSImage:Red Hat Enterprise Linux CoreOS 414.92.202309121329-0 (Plow),ContainerRuntimeVersion:cri-o://1.27.1-7.rhaos4.14.git57d0fe5.el9,KubeletVersion:v1.27.4+2c287eb,KubeProxyVersion:v1.27.4+2c287eb,OperatingSystem:linux,Architecture:amd64,},Images:[]ContainerImage{ContainerImage{Names:[registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-1fllgz2/stable@sha256:f3d1f991c21485c0ce342c15e3364fbd65feaf8c8551dbc0472a187e0711e578],SizeBytes:801282455,},ContainerImage{Names:[registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-1fllgz2/stable@sha256:23012b3380ffce706aa8f204cdc26745d8a69b0218150ec3bcb495202694fdab],SizeBytes:714641214,},ContainerImage{Names:[registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-1fllgz2/stable@sha256:15f1974647ac40b8fc895474cc6db94a553929020a2492732b49e50e89830d74],SizeBytes:498685551,},ContainerImage{Names:[registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-1fllgz2/stable@sha256:0720ec7a4f788ef0327c52d0c2ec732b34f4bb6bb68968d253761ff74394041f],SizeBytes:492764068,},ContainerImage{Names:[registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-1fllgz2/stable@sha256:257eecb1b8a9bfd55fdd8c7c4044398de91241920088f7729fb3303241177e82],SizeBytes:477792737,},ContainerImage{Names:[registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-1fllgz2/stable@sha256:0b996a8dd7422387944eba766203e40f519af120841e1bc5fa533f97c7bd7180],SizeBytes:405326430,},},VolumesInUse:[],VolumesAttached:[]AttachedVolume{},Config:nil,},}
```

kmala replied:
Understood. The reason i am asking is to see if there is another issue because of which we are not setting the provider id in the node. Can you check the kubelet config and logs?

mkowalski replied:
I doubt this is really the case in my environment but from just going over the code I can see that it can happen that we have cloud provider that doesn't use provider ids and only relies on node name, i.e.

```
providerID, err := cloudprovider.GetInstanceProviderID(ctx, cnc.cloud, types.NodeName(node.Name))
if err == cloudprovider.NotImplemented {
	// if the cloud provider being used does not support provider IDs,
	// we can safely continue since we will attempt to set node
	// addresses given the node name in getNodeAddressesByProviderIDOrName
	klog.Warningf("cloud provider does not set node provider ID, using node name to discover node %s", node.Name)
	return "", nil
}
```

in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L580-L603

kmala replied:
got it. If you are using a provider which is open source can you share it. Would like to take a look once.

kmala commented on staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -197,6 +197,7 @@ func (c *CloudNodeLifecycleController) MonitorNodes(ctx context.Context) {
 }
 
 // shutdownInCloudProvider returns true if the node is shutdown on the cloud provider
+// If provider id in spec is empty it calls instanceId with node name to get provider id
```
shutdownInCloudProvider and ensureNodeExistsByProviderID seems to be used only once in the MonitorNodes, can we get the provider and pass it to the function instead of calling twice?

mkowalski replied:
This would be quite some refactor because the whole interface 
```
type InstancesV2 interface {
	InstanceExists(ctx context.Context, node *v1.Node) (bool, error)
	InstanceShutdown(ctx context.Context, node *v1.Node) (bool, error)
	InstanceMetadata(ctx context.Context, node *v1.Node) (*InstanceMetadata, error)
}
```
relies on passing the note object and not just an id. Asked via slack (https://kubernetes.slack.com/archives/C718BPBQ8/p1694687948972739) if that's desired here or not

mkowalski replied:
As discussed, changed only signatures for
```
func shutdownInCloudProvider(ctx context.Context, cloud cloudprovider.Interface, node *v1.Node, providerID string) (bool, error)
```
and
```
func ensureNodeExistsByProviderID(ctx context.Context, cloud cloudprovider.Interface, node *v1.Node, providerID string)
```

mkowalski replied:
And after final discussion, back to the original function signature

andrewsykim commented on staging/src/k8s.io/cloud-provider/fake/fake.go:
```diff
@@ -312,6 +312,11 @@ func (f *Cloud) InstanceExistsByProviderID(ctx context.Context, providerID strin
 	f.addCall("instance-shutdown-by-provider-id")
+
+	if providerID == "" {
+		return false, fmt.Errorf("cannot shutdown instance with empty providerID")
```
Is it possible to add a test case where calling `instances.InstanceID(ctx, types.NodeName(node.Name))` returns no providerID? In this case we would return early and not add the taint

andrewsykim replied:
> I may be missing something but if I read correctly through MonitorNodes it seems to me that node with no providerID would be deleted as it hits the path with

I was referring to InstanceID returning no providerID but not returning the `InstanceNotFound` error. I think we only delete if InstanceID returns InstanceNotFound

mkowalski commented:
/retest-required

All 3 are unrelated failures

kmala commented:
/test pull-kubernetes-e2e-kind-ipv6

kmala commented on staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -150,9 +150,37 @@ func (c *CloudNodeLifecycleController) MonitorNodes(ctx context.Context) {
 
+		// If provider id in spec is empty it calls instanceId with node name to get provider id
+		providerID := node.Spec.ProviderID
+		if providerID == "" {
```
This loop is only needed on  cloud.InstancesV2() is not present. I think we can go back to your original code. Sorry about that.I was just trying to see if we can remove the additional cloud api call.

mkowalski replied:
Sure no problem, after all decided to revert to the very first model

aojea commented:
/retest
> Kubernetes e2e suite: [It] [sig-apps] StatefulSet Basic StatefulSet functionality [StatefulSetBasic] should perform rolling updates and roll backs of template modifications with PVCs expand_less	
> E2eNode Suite: [It] [sig-node] OOMKiller [LinuxOnly] [NodeConformance] init container The containers terminated by OOM killer should have the reason set to OOMKilled expand_more

aojea commented:
/lgtm

@tobiasgiese can you confirm this PR solves the issue

/assign @andrewsykim 

Andrew can we get a final review on this?

tobiasgiese commented on staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -207,7 +208,24 @@ func shutdownInCloudProvider(ctx context.Context, cloud cloudprovider.Interface,
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
+		var err error
+		providerID, err = instances.InstanceID(ctx, types.NodeName(node.Name))
```
I faced the same issue and created a PR for it (already closed it).
Maybe we have to change it here as well, the instance ID is no provider ID and has no provider prefix.
Don't know if it's necessary as the `InstanceShutdownByProviderID` is stripping the provider prefix anyway. But maybe it's worth of adding it, analogous to this in PR https://github.com/kubernetes/kubernetes/pull/121383
https://github.com/kubernetes/kubernetes/blob/8c0af3ee5ba52eefcc61740fa3d94206ea3f4d3c/staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go#L269

mkowalski replied:
How do you want to handle that? I can see what you mean about the diff between provider ID and instance ID, that is valid. At the time when I created this patch I had no idea, but at the same time something along the way makes the assignment I do here a working one (like you suggest, something strips off the prefix later?)

I am happy to add prefix here but I am also happy to keep it as it is given that you tested the patch and also for my environment it fixed the issues I had.

tobiasgiese replied:
Totally fine for me to keep it as is. Maybe we can create a follow issue/PR that fixes all occurrences.

mmerkes replied:
If we want more fine-grained control over fetching the instance ID, we could just add another `getProviderId` method that takes in the `instances.InstanceID` result. Very small change. Seems kind of wrong to build an invalid provider ID when we know that's what we're doing and it's a minor change. But I do agree that fixes all of the other occurrences makes sense as a follow up.

tobiasgiese replied:
Just take a look at my PR https://github.com/kubernetes/kubernetes/pull/121383, I introduced a method `getProviderID` :)

mkowalski replied:
Ok as it's not a big deal I created a function `getProviderID` similarly to the current `ensureNodeExistsByProviderID` and `shutdownInCloudProvider`. It will do what we discussed here, prepending the provider name so we are building a correct provider ID.

I would appreciate a look regarding the instanceV2 handling inside `getProviderID`. Today it is not used because `shutdownInCloudProvider` handles instanceV2 interface separately and will never call `getProviderID` but I still want to handle it in case someone somewhere start using `getProviderID` (because why not if it's convenient)

mmerkes replied:
nit: `providerID, err = instances.InstanceID(ctx, types.NodeName(node.Name))` shouldn't `providerID` be `instanceID` in this case?

mkowalski replied:
Mhm, of course that makes sense. I thought I could piggyback on having only single variable used all around but it's not a good idea

tobiasgiese commented:
> @tobiasgiese can you confirm this PR solves the issue

I can confirm that this PR solves the issue. I also tested it in our test environment and could reproduce the issue with 30 worker nodes and with this change the issue is fixed.

aojea commented:
/priority important-soon

tobiasgiese commented on staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -196,7 +196,50 @@ func (c *CloudNodeLifecycleController) MonitorNodes(ctx context.Context) {
+// it will be the one from the cloud provider.
+func getProviderID(ctx context.Context, cloud cloudprovider.Interface, node *v1.Node) (string, error) {
+	var providerID string
+	var err error
```
You don't have to initialize the var `err`

mkowalski replied:
Correct me where I am wrong but if I don't declare it here then I need to write
```
providerID, err := instances.InstanceID(ctx, types.NodeName(node.Name))
```
what is going to shadow providerID and subsequently ignore the result of
```
instances.InstanceID(ctx, types.NodeName(node.Name))
```
when used in
```
return cloud.ProviderName() + "://" + providerID, nil
```

tobiasgiese replied:
You are totally right, I have missed the `providerID` var 🤦🏻

tobiasgiese commented on staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -207,7 +250,12 @@ func shutdownInCloudProvider(ctx context.Context, cloud cloudprovider.Interface,
 	}
 
-	shutdown, err := instances.InstanceShutdownByProviderID(ctx, node.Spec.ProviderID)
+	providerID, err := getProviderID(ctx, cloud, node)
```
You can adjust `ensureNodeExistsByProviderID` as well, it's the same logic

tobiasgiese commented:
/lgtm

mkowalski commented:
/test pull-kubernetes-integration

I hope/believe/guess that it's just a flake

mmerkes commented:
/lgtm

tobiasgiese commented:
/lgtm

tobiasgiese commented on staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -196,6 +196,47 @@ func (c *CloudNodeLifecycleController) MonitorNodes(ctx context.Context) {
+	}
+
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
```
One last nit: We should return early here

```suggestion
	if providerID != "" {
		return providerID, nil
	}
```

tobiasgiese replied:
Also regarding [Code Health: Reduce Nesting, Reduce Complexity](https://testing.googleblog.com/2017/06/code-health-reduce-nesting-reduce.html)

mkowalski replied:
Good hints indeed, thank you, just fixed the code!

mmerkes commented:
/test pull-kubernetes-e2e-gce

aojea commented on staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -216,7 +263,6 @@ func shutdownInCloudProvider(ctx context.Context, cloud cloudprovider.Interface,
 // ensureNodeExistsByProviderID checks if the instance exists by the provider id,
-// If provider id in spec is empty it calls instanceId with node name to get provider id
 func ensureNodeExistsByProviderID(ctx context.Context, cloud cloudprovider.Interface, node *v1.Node) (bool, error) {
 	if instanceV2, ok := cloud.InstancesV2(); ok {
```
this is called twice now, here and inside `getProviderID`

mkowalski replied:
To save on this `InstancesV2` call I could modify the signature of all 3 functions to be
```
func chocobomb(ctx context.Context, cloud cloudprovider.Interface, node *v1.Node, instancesV2 bool)
```
so that in the whole `MonitorNodes` we would call it only once. That would be even better than what we have today because today we call it twice.

Similarly with `cloud.Instances()`, it could be passed as part of signature e.g.

```
func chocobomb(ctx context.Context, cloud cloudprovider.Interface, node *v1.Node, instancesV1 bool, instancesV2 bool)
```

and again we optimized from 2 calls today to 1 call tomorrow. How do you feel about such an approach?

andrewsykim replied:
Calling `cloud.InstancesV2()` twice is not a big deal, it usually just returns a field to a struct

aojea commented on staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -196,6 +196,48 @@ func (c *CloudNodeLifecycleController) MonitorNodes(ctx context.Context) {
+// getProviderID returns the provider ID for the node. If Node CR has no provider ID,
+// it will be the one from the cloud provider.
+func getProviderID(ctx context.Context, cloud cloudprovider.Interface, node *v1.Node) (string, error) {
+	if instanceV2, ok := cloud.InstancesV2(); ok {
+		metadata, err := instanceV2.InstanceMetadata(ctx, node)
+		if err != nil {
+			if err == cloudprovider.InstanceNotFound {
+				return "", nil
+			}
+			return "", err
+		}
+
+		return metadata.ProviderID, nil
+	}
+
+	instances, ok := cloud.Instances()
+	if !ok {
+		return "", errors.New("cloud provider does not support instances")
+	}
```
this code is called before in the two functions that call this one, why do we need to duplicate the cloud provider calls?

mkowalski replied:
I did see it when I wrote it but I had a trouble simply removing that call... The way I think is that I have `getProviderID` which takes `cloud cloudprovider.Interface` as one of parameters. Sure, it's true that `shutdownInCloudProvider` and `ensureNodeExistsByProviderID` call `cloud.InstancesV2()` already but what if someone introduces a 3rd function that uses `getProviderID` somewhere in its body but doesn't check for V2 before? 

Today, with only 2 consumers, we have guarantee that `getProviderID` is only called for non-V2. But tomorrow, how to ensure this is the case? If you have any suggestion I am more than happy to hear!

aojea replied:
this is a DRY vs WET example to me https://www.zippyops.com/is-your-code-dry-or-wet#:~:text=DRY%20code%20is%20a%20software,t%20adhere%20to%20DRY%20principle., just repeat the code twice, if there ia third function that wants to use it that is the moment to consolidate it

My 2 cents

mkowalski replied:
It is tempting to remove both `if instanceV2, ok := cloud.InstancesV2();` and `instances, ok := cloud.Instances()` and simply say that whoever calls `getProviderID` needs to make sure that those prerequisites are met. It feels okay because all those helper functions in this file are local and no one else uses them. It feels wrong because they can very easily become generic utility functions used somewhere else and break stuff if someone abuses them. Let's hear people out, it's hard for me to have a strong opinion about any possible approach here

tobiasgiese commented on staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -196,18 +196,49 @@ func (c *CloudNodeLifecycleController) MonitorNodes(ctx context.Context) {
+		return metadata.ProviderID, nil
+	}
+
+	providerID, err := cloudprovider.GetInstanceProviderID(ctx, c.cloud, types.NodeName(node.Name))
```
With your implementation we cannot return the correct value if `cloudprovider.InstanceNotFound`.
Please take a look at my implementation in PR https://github.com/kubernetes/kubernetes/pull/121383, I had considered everything there IMO

mkowalski replied:
Yeah I noticed this, I kind of ignored this because I feel this will be rewritten anyway (look at my bigger comment below) but I will definitely keep that in mind to handle `cloudprovider.InstanceNotFound` once we agree on the high-level shape of things

mkowalski commented:
To somehow sum up the discussion because it keeps disappearing with the code changes... In the today's form of the `node_lifecycle_controller.go` we have a nested chain of calls to the `cloud.InstancesV2()`, i.e.

* MonitorNodes
    * ensureNodeExistsByProviderID <-- 1 call to cloud.InstancesV2(), 1 call to cloud.Instances()
        * getProviderID <-- 1 call to cloud.InstancesV2()
            * cloudprovider.GetInstanceProviderID <-- 1 call to cloud.Instances()
    * shutdownInCloudProvider <-- 1 call to cloud.InstancesV2(), 1 call to cloud.Instances()
        * getProviderID <-- 1 call to cloud.InstancesV2()
            * cloudprovider.GetInstanceProviderID <-- 1 call to cloud.Instances()

so a single iteration over `MonitorNodes` gives us 4 calls to both `InstancesV2()` and `Instances()`. This is suboptimal, I think that's clear.

I did have a look at `staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go` because that's a controller that also uses InstancesV2 interface and sadly similar pattern occurs there. I do not have a silver bullet to optimize it, only some skeletons of ideas. I would be welcome to discuss this further.

---

Something that pops up to my mind is the controller struct we have today,

```
type CloudNodeLifecycleController struct {
	kubeClient clientset.Interface
	nodeLister v1lister.NodeLister

	broadcaster record.EventBroadcaster
	recorder    record.EventRecorder

	cloud cloudprovider.Interface

	// Value controlling NodeController monitoring period, i.e. how often does NodeController
	// check node status posted from kubelet. This value should be lower than nodeMonitorGracePeriod
	// set in controller-manager
	nodeMonitorPeriod time.Duration
}
```

and I am asking if we could extend it to e.g.

```
type CloudNodeLifecycleController struct {
	kubeClient clientset.Interface
	nodeLister v1lister.NodeLister

	broadcaster record.EventBroadcaster
	recorder    record.EventRecorder

	cloud cloudprovider.Interface

        instanceV1 cloudprovider.Instances
        instanceV2 cloudprovider.InstancesV2

	// Value controlling NodeController monitoring period, i.e. how often does NodeController
	// check node status posted from kubelet. This value should be lower than nodeMonitorGracePeriod
	// set in controller-manager
	nodeMonitorPeriod time.Duration
}
```

so that when we initialize it in `NewCloudNodeLifecycleController` we would already there set the corresponding `instance{V1|V2}` field. The function already calls the what's needed so we wouldn't even add a single call there

```
	_, instancesSupported := cloud.Instances()
	_, instancesV2Supported := cloud.InstancesV2()
	if !instancesSupported && !instancesV2Supported {
		return nil, errors.New("cloud provider does not support instances")
	}

	c := &CloudNodeLifecycleController{
		kubeClient:        kubeClient,
		nodeLister:        nodeInformer.Lister(),
		broadcaster:       eventBroadcaster,
		recorder:          recorder,
		cloud:             cloud,
		nodeMonitorPeriod: nodeMonitorPeriod,
	}
```

What I am not sure is whether keeping `cloud.Instances` at the level of controller is not going to introduce some race somewhere. Happy to hear pros and cons

mkowalski committed:
gce: fix test for non-existing instance in cloud

This commit fixes a GCE instance test for a scenario where instance does
not exist in the cloud. In the `gce_instances.go` we have a code that
handles such a scenario and strips off the error, i.e.

```
if providerID, err = cloudprovider.GetInstanceProviderID(ctx, g, types.NodeName(node.Name)); err != nil {
	if err == cloudprovider.InstanceNotFound {
		return false, nil
	}
	return false, err
}
```

but nevertheless the test was expecing a non-empty error string. This
issue got exposed when implementation of `GetInstanceProviderID` in the
`cloud-provider/cloud.go` changed to return `InstanceNotFound` as-is and
not as a custom string.

Signed-off-by: Mat Kowalski <mko@redhat.com>

mkowalski commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

aojea commented on staging/src/k8s.io/cloud-provider/cloud.go:
```diff
@@ -108,6 +108,9 @@ func GetInstanceProviderID(ctx context.Context, cloud Interface, nodeName types.
+			return "", err
+		}
 
 		return "", fmt.Errorf("failed to get instance ID from cloud provider: %v", err)
```
Since we are doing this it seems more appropiate to wrap the error so we can not keep growing the if clauses
```suggestion
		return "", fmt.Errorf("failed to get instance ID from cloud provider: %w", err)
```

mkowalski replied:
Sure, no problem

mkowalski commented:
/test pull-kubernetes-node-e2e-containerd

andrewsykim commented on staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -196,18 +196,52 @@ func (c *CloudNodeLifecycleController) MonitorNodes(ctx context.Context) {
+	providerID, err := cloudprovider.GetInstanceProviderID(ctx, c.cloud, types.NodeName(node.Name))
+	if err != nil {
+		if err == cloudprovider.InstanceNotFound {
+			return "", nil
```
instead of returning nil here, can we return `"", err` and check for `cloudprovider.InstanceNotFound` from  `ensureNodeExistsByProviderID` and `shutdownInCloudProvider`?

mkowalski replied:
I have nothing against, it's just the code duplication that was pointed out yesterday and I tried to remove. The next time we have a function that uses `getProviderID` it will need to again have the `if err == cloudprovider.InstanceNotFound`

andrewsykim commented on staging/src/k8s.io/cloud-provider/cloud.go:
```diff
@@ -108,8 +108,11 @@ func GetInstanceProviderID(ctx context.Context, cloud Interface, nodeName types.
 			return "", err
 		}
+		if err == InstanceNotFound {
+			return "", err
```
historically `GetInstanceProviderID` never returned InstanceNotFound and we only expected it form InstanceID. I think this change is correct but we should update the comments for GetInstanceProviderID to reflect this change.

andrewsykim replied:
See https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L181 for example

mkowalski replied:
Before my change `GetInstanceProviderID` was returning a string
```
"failed to get instance ID from cloud provider: instance not found"
```
and with my change it returns
```
"instance not found"
```

Given that there is `InstanceNotFound = errors.New("instance not found")` defined I felt it's a right thing to do. I will update the comments accordingly

andrewsykim replied:
The change is fine I think, but keep in mind that some callers may be checking for the `InstanceNotFound` error.

andrewsykim commented on staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -216,32 +250,19 @@ func shutdownInCloudProvider(ctx context.Context, cloud cloudprovider.Interface,
-		}
+	providerID, err := c.getProviderID(ctx, node)
+	if err != nil {
+		return false, err
```
`ensureNodeExistsByProviderID` needs to return `false, nil` if instance doesn't exist. In this case it would run `instances.InstanceExistsByProviderID(ctx, providerID)` on line 268 with an empty providerID resulting in an error because `getProviderID` returns `"", nil`.

andrewsykim commented on staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:
```diff
@@ -196,18 +196,52 @@ func (c *CloudNodeLifecycleController) MonitorNodes(ctx context.Context) {
+	if instanceV2, ok := c.cloud.InstancesV2(); ok {
+		metadata, err := instanceV2.InstanceMetadata(ctx, node)
+		if err != nil {
+			if err == cloudprovider.InstanceNotFound {
```
Same comment here, can we just return `"", err` here instead of nil when err is `cloudprovider.InstanceNotFound`?

andrewsykim reviewed (commented):
/approve

@aojea can you do another pass and the final LGTM?

aojea commented:
/lgtm

Thanks

pacoxu commented:
/retest
for a flake

aojea commented:
@mkowalski do you want to backport this for all the stable branches?

mkowalski commented:
> @mkowalski do you want to backport this for all the supported branches?

Definitely yes, I will prepare backports later today
